crates/ruma-client-api/src/room/get_event_by_timestamp.rs

```diff
@@ -0,0 +1,63 @@
+//! `GET /_matrix/client/*/rooms/{roomId}/timestamp_to_event`
+//!
+//! Get the ID of the event closest to the given timestamp.
+
+pub mod unstable {
+    //! `/unstable/` ([spec])
+    //!
+    //! [spec]: https://github.com/matrix-org/matrix-spec-proposals/pull/3030
+
+    use ruma_common::{
+        api::{request, response, Direction, Metadata},
+        metadata, MilliSecondsSinceUnixEpoch, OwnedEventId, OwnedRoomId,
+    };
+
+    const METADATA: Metadata = metadata! {
+        method: GET,
+        rate_limited: true,
+        authentication: AccessToken,
+        history: {
+            unstable => "/_matrix/client/unstable/org.matrix.msc3030/rooms/:room_id/timestamp_to_event",
+        }
+    };
+
+    /// Request type for the `get_event_by_timestamp` endpoint.
+    #[request(error = crate::Error)]
+    pub struct Request {
+        /// The ID of the room the event is in.
+        #[ruma_api(path)]
+        pub room_id: OwnedRoomId,
+
+        /// The timestamp to search from.
+        #[ruma_api(query)]
+        pub ts: MilliSecondsSinceUnixEpoch,
+
+        /// The direction in which to search.
+        #[ruma_api(query)]
+        pub dir: Direction,
+    }
+
+    /// Response type for the `get_event_by_timestamp` endpoint.
+    #[response(error = crate::Error)]
+    pub struct Response {
+        /// The ID of the event found.
+        pub event_id: OwnedEventId,
+
+        /// The event's timestamp.
+        pub origin_server_ts: MilliSecondsSinceUnixEpoch,
+    }
+
+    impl Request {
```

Comment:
How about adding constructors where you don't have to pass the `Direction`? I think we could call them `before` and `after` if the timestamp has exclusive semantics or `until` and `since` if it has inclusive semantics (would also be nice to clarify this in the field's docs).